Assets/BossRoom/Scripts/Client/UI/Lobby/LobbyJoiningUI.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Collections.Generic;
+using BossRoom.Scripts.Shared.Infrastructure;
+using BossRoom.Scripts.Shared.Net.UnityServices.Lobbies;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace BossRoom.Scripts.Client.UI
+{
+    /// <summary>
+    /// Handles the list of LobbyListItemUIs and ensures it stays synchronized with the lobby list from the service.
+    /// </summary>
+    public class LobbyJoiningUI : MonoBehaviour
+    {
+        [SerializeField] private LobbyListItemUI m_LobbyListItemPrototype;
+        [SerializeField] private InputField m_JoinCodeField;
+        [SerializeField] private CanvasGroup m_CanvasGroup;
+
+
+        private IInstanceResolver m_Container;
+
+        private LobbyUIMediator m_LobbyUIMediator;
+
+        private UpdateRunner m_UpdateRunner;
+
+        private IDisposable m_Subscriptions;
+
+        private List<LobbyListItemUI> m_LobbyListItems = new List<LobbyListItemUI>();
+
+        private void Awake()
+        {
+            m_LobbyListItemPrototype.gameObject.SetActive(false);
+        }
+
+        private void OnDisable()
+        {
+            if (m_UpdateRunner != null)
+            {
+                m_UpdateRunner.Unsubscribe(PeriodicRefresh);
+            }
+        }
+
+        private void OnDestroy()
+        {
+            m_Subscriptions?.Dispose();
+        }
+
+        [Inject]
+        private void InjectDependenciesAndInstantiate(
+            IInstanceResolver container,
+            LobbyUIMediator lobbyUIMediator,
+            UpdateRunner updateRunner,
+            ISubscriber<LobbyListFetchedMessage> localLobbiesRefreshedSub)
+        {
+            m_Container = container;
+            m_LobbyUIMediator = lobbyUIMediator;
+            m_UpdateRunner = updateRunner;
+
+            m_UpdateRunner.Subscribe(PeriodicRefresh, 10f);
+
+            m_Subscriptions = localLobbiesRefreshedSub.Subscribe(UpdateUI);
+        }
+
+        public void OnJoinButtonPressed()
+        {
+            m_LobbyUIMediator.JoinLobbyWithCodeRequest(m_JoinCodeField.text.ToUpper());
+        }
+
+        private void PeriodicRefresh(float _)
+        {
+            m_LobbyUIMediator.QueryLobbiesRequest(false);
+        }
+
+        public void OnRefresh()
+        {
+            m_LobbyUIMediator.QueryLobbiesRequest(true);
```

Comment:
Should the local list be cleared here while waiting for the refresh results? Or when this window is closed? When testing, if there is a lobby on and I join it, then shut down the session with the host and reopen the window, the client will briefly see this old lobby still active until it is refreshed. This feels a bit weird.

Reply:
Also, if I spam the refresh button, I get rate-limited quite quickly, but if I continue I end up with the game not responding

Reply:
It shouldn't be cleared while we are just refreshing the lobbies - they might just get their state updated (such as when an extra player joins), or they can go away entirely (when Relay gets a disconnect from the host, it will end the lobby eventually, however we are also sending a quit lobby message on leave or app quit).

Reply:
The worst that would happen is that you somehow manage to click a button for a lobby that's already dead on the server, but we didn't fetch that data yet - then you just won't be able to join and the lobby will go away eventually.

Reply:
In that case, should we hide it while it refreshes? In the end it's pretty subjective but I feel it makes sense to not see the list of lobbies while it refreshes. If not, I think it could be cleared when closing the window to avoid the case I described? It's really not a big deal though, just something that feels weird to me

Reply:
Concerning the button spam leading to the game not responding, it is worse on release builds as the log doesn't appear and we don't have a popup telling us that we are getting rate limited

Reply:
It's really the whole application that is not responding. I left it a few minutes like that and it didn't unfreeze, I had to force it to close itself. Reproduced in editor and with a release and a dev build. To reproduce, simply spam the refresh button until you get rate limited, then continue spamming it as soon as the button is re-enabled. After around 7-8 clicks the game freezes and stops responding

Reply:
should be fixed.